core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategy.java

```diff
@@ -112,11 +114,49 @@
      * Returns a {@link CompletionStage} that contains {@link Backoff} which will be used for retry.
      * If the condition does not match, this will return {@code null} to stop retry attempt.
      * Note that {@link ResponseTimeoutException} is not retriable for the whole retry, but each attempt.
+     * To retrieve the response {@link HttpHeaders} without {@link Response}, you can use the specified
+     * {@link ClientRequestContext}:
+     *
+     * <pre>{@code
+     * > CompletionStage<Backoff> shouldRetry(ClientRequestContext ctx, @Nullable T response,
+     * >                                      @Nullable Throwable cause) {
+     * >     if (response == null) {
+     * >         HttpHeaders responseHeaders = ctx.log().responseHeaders();
+     * >         if (responseHeaders.status().codeClass() == HttpStatusClass.SERVER_ERROR) {
+     * >             return CompletableFuture.completedFuture(backoff);
+     * >         }
+     * >     }
+     * >     ...
+     * > }
+     * }</pre>
+     *
+     * @param ctx the {@link ClientRequestContext} of this request
+     * @param response the {@link Response} from the server. {@code null} if
+     *                 {@link #needsResponseStreaming(ClientRequestContext)} returns {@code false} and
+     *                 the type of {@link Response} is {@link HttpResponse}. The actual {@link Response}
+     *                 otherwise.
+     * @param cause the {@link Throwable} which is raised while sending a request. {@code null} it there's no
+     *              exception.
      *
      * @see RetryingClientBuilder#responseTimeoutMillisForEachAttempt(long)
      *
      * @see <a href="https://line.github.io/armeria/advanced-retry.html#per-attempt-timeout">Per-attempt
      *      timeout</a>
      */
-    CompletionStage<Backoff> shouldRetry(I request, O response);
+    CompletionStage<Backoff> shouldRetry(ClientRequestContext ctx, @Nullable T response,
+                                         @Nullable Throwable cause);
+
+    /**
+     * Returns {@code true} if you need to determine whether you need to retry by looking into the
+     * {@link HttpResponse} content in {@link #shouldRetry(ClientRequestContext, Response, Throwable)}.
+     * That will invoke {@link #shouldRetry(ClientRequestContext, Response, Throwable)} with the actual
```

Comment:
`If this method returns {@code true}, {@link #shouldRetry...} will be invoked with ...`